src/main/java/org/reaktivity/nukleus/amqp/internal/stream/AmqpServerFactory.java

```diff
@@ -2698,6 +2726,8 @@ private void onDecodeTransfer(
                     sessions.remove(incomingChannel);
                     flushReplySharedBudget(traceId);
                     doEncodeEnd(traceId, authorization, outgoingChannel, SESSION_WINDOW_VIOLATION);
+                    sessionState = sessionState.sentEnd();
+                    assert sessionState == AmqpSessionState.DISCARDING;
```

Comment:
Could these be moved to a session-scoped `doEncodeEnd(traceId, authorization)` with fewer arguments instead?
Then it doesn't matter why `END` was sent, the state machine always represents the session state accurately.

Reply:
Can I add error type argument in `doEncodeEnd(traceId, authorization)`?  
```
            private void doEncodeEnd(
                long traceId,
                long authorization,
                AmqpErrorType errorType)
            {
                cleanup(traceId, authorization);
                sessions.remove(incomingChannel);
                flushReplySharedBudget(traceId);
                AmqpServer.this.doEncodeEnd(traceId, authorization, outgoingChannel, errorType);
                sessionState = sessionState.sentEnd();
                assert sessionState == AmqpSessionState.DISCARDING;
            }
```

Reply:
Yes, of course!